compiler-plugin/src/main/kotlin/com/google/devtools/ksp/Incremental.kt

```diff
@@ -428,6 +453,129 @@ class IncrementalContext(
         updateOutputs(outputs, cleanOutputs)
 
     }
+
+    fun recordLookup(psiFile: PsiJavaFile, fqn: String) {
+        val path = psiFile.virtualFile.path
+        val name = fqn.substringAfterLast('.')
+        val scope = fqn.substringBeforeLast('.', "<anonymous>")
+
+        fun record(scope: String, name: String) =
+            lookupTracker.record(path, Position.NO_POSITION, scope, ScopeKind.CLASSIFIER, name)
+
+        record(scope, name)
+        val onDemandImports =
+                psiFile.getOnDemandImports(false, false).mapNotNull { (it as? PsiPackage)?.qualifiedName }
+        if (scope in onDemandImports) {
+            record(psiFile.packageName, name)
+            onDemandImports.forEach {
+                record(it, name)
+            }
+        }
+    }
+
+    private fun recordLookup(ref: PsiClassReferenceType, def: PsiClass) {
+        val psiFile = ref.reference.containingFile as? PsiJavaFile ?: return
+        // A type parameter doesn't have qualified name.
```

Comment:
Looks like we just don't record for type parameters, what should be the ideal behavior for a type parameter has an upper bound which is concrete?

Reply:
We don't do it here. This function is just used to record a type reference to a definition. Type parameters' bounds are recorded in other places such as `recordLookup(psi: PsiMethod)` only when necessary.

Will add a comment.